core/modules/qana/QservRestrictorPlugin.cc

```diff
@@ -571,14 +574,36 @@ QservRestrictorPlugin::_newSecIndexRestrictor(
         return query::QsRestrictor::Ptr();
     }
 
+    // Get the director table (or an empty string)
+    std::string dirTable = context.cssFacade->getDirTable(cr->db, cr->table);
+
     // Build the QsRestrictor
     query::QsRestrictor::Ptr p = std::make_shared<query::QsRestrictor>();
     p->_name = "sIndex";
     // sIndex has paramers as follows:
     // db, table, column, val1, val2, ...
-    p->_params.push_back(cr->db);
-    p->_params.push_back(cr->table);
-    p->_params.push_back(cr->column);
+    if (!dirTable.empty()) {
+        std::string dirDb = context.cssFacade->getDirDb(cr->db, cr->table);
+        if (dirDb.empty()) {
+            dirDb = cr->db;
+        }
+        std::string dirCol = context.cssFacade->getDirColName(cr->db, cr->table);
+        if (dirCol.empty()) {
+            LOGF_ERROR("dirTable set, dirCol missing for %1%.%2%" % cr->db % cr->table);
+            return query::QsRestrictor::Ptr();
+        }
+        LOGF_DEBUG("Using dirDb %1%, dirTable %2%, dirCol %3% as sIndex for %4%.%5%." %
+                   dirDb % dirTable % dirCol % cr->db % cr->table);
+        p->_params.push_back(dirDb);
+        p->_params.push_back(dirTable);
+        p->_params.push_back(dirCol);
+    } else {
+        LOGF_DEBUG("Using %1%, %2%, %3% as sIndex." %
+                   cr->db % cr->table % cr->column);
+        p->_params.push_back(cr->db);
+        p->_params.push_back(cr->table);
+        p->_params.push_back(cr->column);
+    }
```

Comment:
Hmm... not really! The block

    if (dirDb != cr->db || dirTable != cr->table) {
        // Lookup the name of the director column in the director table
        dirCol = context.cssFacade->getDirColName(dirDb, dirTable);
        if (dirCol.empty()) {
            LOGF_ERROR("dirCol missing for %1%.%2%" % dirDb % dirTable);    
            return query::QsRestrictor::Ptr();
        }
    }

is not needed because we are setting dirCol earlier, so it is guaranteed to be set.

But most importantly, I always thought that if we set dirColName, we must set dirTable as well (and I assume just-introduced dirDb can be left out, in which case it defaults to current db). Does it really make sense to set dirColName without dirTable? If you agree that dirColName and dirTable always go together, than we need to rewrite it. How about:

    // Build the QsRestrictor
    query::QsRestrictor::Ptr p = std::make_shared<query::QsRestrictor>();
    p->_name = "sIndex";
    // sIndex has paramers as follows:
    // db, table, column, val1, val2, ...

    // Get the director db/tb/col info
    std::string dirDb = context.cssFacade->getDirDb(cr->db, cr->table);
    std::string dirTable = context.cssFacade->getDirTable(cr->db, cr->table);
    std::string dirCol = context.cssFacade->getDirColName(cr->db, cr->table);
    if (dirDb.empty() && dirTable.empty() && dirCol.empty()) {
        LOGF_DEBUG("Using %1%, %2%, %3% as sIndex." % cr->db % cr->table % cr->column);
        p->_params.push_back(cr->db);
        p->_params.push_back(cr->table);
        p->_params.push_back(cr->column);
    } else {
        if (dirTable.empty() && !dirCol.empty() ) {
            LOGF_ERROR("dirCol missing, dirTable set to %1%" % dirTable);
            return query::QsRestrictor::Ptr();
        }
        if (dirCol.empty() && !dirTable.empty() ) {
            LOGF_ERROR("dirTable missing, dirCol set to %1%" % dirCol);
            return query::QsRestrictor::Ptr();
        }
        if (!dirDb.empty() && (dirTable.empty() or dirCol.empty()) {
            LOGF_ERROR("dirTable/dirCol missing, dirDb set to %1%" % dirDb);
            return query::QsRestrictor::Ptr();
        }
        if (dirDb.empty()) {
            dirDb = cr->db;
        }
        if (dirCol != cr->col) {
            LOGF_ERROR("dirCol=%1 set inconsistently for %2%.%3%.%4%" % \
		       dirColcr->db % cr->table % cr->col);
            return query::QsRestrictor::Ptr();
        }
        p->_params.push_back(dirDb);
        p->_params.push_back(dirTable);
        p->_params.push_back(dirCol);
    }